src/MongoDB.Driver.Core/Core/Misc/ServerVersion.cs

```diff
@@ -0,0 +1,463 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace MongoDB.Driver.Core.Misc
+{
+    internal class ServerVersion : IEquatable<ServerVersion>, IComparable<ServerVersion>
+    {
+        // fields
+        private readonly string _commitHash;
+        private readonly int? _internalBuild;
+        private readonly int _major;
+        private readonly int _minor;
+        private readonly int _patch;
+        private readonly int? _releaseCandidate;
+
+        // constructors
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ServerVersion"/> class.
+        /// </summary>
+        /// <param name="major">The major version.</param>
+        /// <param name="minor">The minor version.</param>
+        /// <param name="patch">The patch version.</param>
+        public ServerVersion(int major, int minor, int patch)
+            : this(major, minor, patch, null)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ServerVersion"/> class.
+        /// </summary>
+        /// <param name="major">The major version.</param>
+        /// <param name="minor">The minor version.</param>
+        /// <param name="patch">The patch version.</param>
+        /// <param name="releaseCandidate">The release candidate version.</param>
+        public ServerVersion(int major, int minor, int patch, int? releaseCandidate)
```

Comment:
since we have different ctors with different arguments, I think `releaseCandidate` should be not nullable. But it looks like it will make code a bit more difficult. So, it's up to you